File: src/langarian/metrics.py
"""Metric functions for the finite-dimensional Langarian kernel."""

from __future__ import annotations

from collections.abc import Sequence
import math

import numpy as np

from .limits import MAX_STATES, LimitError, MetricError
from .state import ResonantState, pad_to_common_dim
from .version import METRIC_VERSION  # compatibility re-export; single source is version.py

__all__ = [
    "METRIC_VERSION",
    "MetricError",
    "resonance",
    "phase",
    "normalized_complex_similarity",
    "system_coherence",
]


def resonance(state: ResonantState) -> float:
    """Resonance is vector norm."""

    return state.resonance


def phase(state: ResonantState) -> float:
    """Return the kernel's global phase convention in radians."""

    return state.phase


def normalized_complex_similarity(a: ResonantState, b: ResonantState) -> float:
    """Squared normalized inner-product similarity in [0, 1].

    Definition (exact arithmetic): C(a, b) = |<a, b>|^2 / (||a||^2 * ||b||^2),
    bounded in [0, 1] by the Cauchy-Schwarz inequality. This is the standard
    projective (ray) similarity and is invariant under any nonzero complex
    rescaling of either argument.

    Zero-state convention:
    - C(0, 0) = 1
    - C(0, x) = 0 for nonzero x

    Scale-safety (metric:v0.3): each vector is normalized by its max component
    magnitude before the inner product, so squared norms can neither underflow
    nor overflow. The value is mathematically unchanged in exact arithmetic;
    the zero-state conventions above hold for every finite float64 magnitude,
    including deep subnormals down to 5e-324. When ``maxabs * maxabs``
    underflows to 0 (deep-subnormal ``maxabs``, below ~2.3e-162), numpy
    complex scalar division divides by 0 and would produce NaN, so the real
    and imaginary parts are scaled separately instead (exact). A non-finite
    or non-positive intermediate is an explicit MetricError, never a silently
    clamped NaN.
    """

    av, bv = pad_to_common_dim(a, b)
    maxabs_a = float(np.max(np.abs(av)))
    maxabs_b = float(np.max(np.abs(bv)))
    a_is_zero = maxabs_a == 0.0
    b_is_zero = maxabs_b == 0.0
    if a_is_zero and b_is_zero:
        return 1.0
    if a_is_zero or b_is_zero:
        return 0.0
    # Scale into the unit ball: |component| <= 1, so norms and inner products
    # are finite and well-conditioned for any finite input magnitude. Complex
    # scalar division is used when its c*c + d*d denominator is nonzero; for
    # deep-subnormal maxabs it underflows to 0, so real/imaginary parts are
    # scaled separately (exact; see docstring).
    sa = av / maxabs_a if maxabs_a * maxabs_a > 0.0 else (av.real / maxabs_a) + 1j * (av.imag / maxabs_a)
    sb = bv / maxabs_b if maxabs_b * maxabs_b > 0.0 else (bv.real / maxabs_b) + 1j * (bv.imag / maxabs_b)
    inner = np.vdot(sa, sb)
    na2 = float(np.vdot(sa, sa).real)
    nb2 = float(np.vdot(sb, sb).real)
    if na2 <= 0.0 or nb2 <= 0.0:
        raise MetricError("scaled squared norm is non-positive for a nonzero state; refusing to clamp.")
    value = (abs(inner) ** 2) / (na2 * nb2)
    if not math.isfinite(value):
        raise MetricError(f"similarity intermediate is non-finite ({value!r}); refusing to clamp NaN/inf.")
    return float(min(1.0, max(0.0, value)))


def system_coherence(states: Sequence[ResonantState], weights: np.ndarray | None = None) -> float:
    """Average pairwise coherence for a finite state system.

    Convention: the diagonal (self-similarities, each equal to 1 by the
    zero-state convention) IS included in the weighted average. This is a
    deliberate averaging convention, not a mathematical requirement.

    Weights must be finite and non-negative; negative weights could push the
    result outside [0, 1] and are rejected with a typed error.
    """

    n = len(states)
    if n == 0:
        raise ValueError("system_coherence requires at least one state.")
    if n > MAX_STATES:
        raise LimitError(f"system_coherence received {n} states; limit is MAX_STATES={MAX_STATES}.")
    if weights is None:
        weights = np.ones((n, n), dtype=float)
    weights = np.asarray(weights, dtype=float)
    if weights.shape != (n, n):
        raise ValueError("weights must have shape (n, n).")
    if not np.all(np.isfinite(weights)):
        raise ValueError("weights must be finite.")
    if np.any(weights < 0):
        raise ValueError("weights must be non-negative; negative weights are not meaningful for coherence averaging.")
    total = 0.0
    weight_total = 0.0
    for i in range(n):
        for j in range(n):
            w = float(weights[i, j])
            total += normalized_complex_similarity(states[i], states[j]) * w
            weight_total += w
    if math.isclose(weight_total, 0.0):
        raise ValueError("weights must not sum to zero.")
    result = float(total / weight_total)
    if not math.isfinite(result):
        raise MetricError(f"system_coherence produced a non-finite value ({result!r}).")
    return result

File: web/src/kernel/metrics.ts
/**
 * Metric functions for the finite-dimensional Langarian kernel (TS mirror).
 *
 * Mirrors src/langarian/metrics.py: scale-safe normalized similarity with
 * the zero conventions C(0,0)=1, C(0,x)=0, and system coherence with
 * non-negative finite weights and diagonal self-similarities included.
 */

import { absCx, ddot, divVectorReal, padToCommonDim, vdot } from './complex.js'
import { MAX_STATES, LimitError, MetricError, ValueError } from './limits.js'
import { ResonantState } from './state.js'

export function resonance(state: ResonantState): number {
  return state.resonance
}

export function phase(state: ResonantState): number {
  return state.phase
}

/**
 * Squared normalized inner-product similarity in [0, 1]:
 * C(a, b) = |<a, b>|^2 / (||a||^2 * ||b||^2).
 *
 * Zero-state convention: C(0, 0) = 1, C(0, x) = 0 for nonzero x.
 * Scale-safety (metric:v0.3): each vector is normalized by its max component
 * magnitude before the inner product. A non-finite intermediate is an
 * explicit MetricError, never a silently clamped NaN.
 */
export function normalizedComplexSimilarity(a: ResonantState, b: ResonantState): number {
  const [av, bv] = padToCommonDim(a.vector, b.vector)
  let maxabsA = 0
  let maxabsB = 0
  for (const z of av) {
    const m = absCx(z)
    if (m > maxabsA) maxabsA = m
  }
  for (const z of bv) {
    const m = absCx(z)
    if (m > maxabsB) maxabsB = m
  }
  const aIsZero = maxabsA === 0
  const bIsZero = maxabsB === 0
  if (aIsZero && bIsZero) return 1
  if (aIsZero || bIsZero) return 0
  const sa = divVectorReal(av, maxabsA)
  const sb = divVectorReal(bv, maxabsB)
  const inner = vdot(sa, sb)
  const saRe = sa.map((z) => z.re)
  const saIm = sa.map((z) => z.im)
  const sbRe = sb.map((z) => z.re)
  const sbIm = sb.map((z) => z.im)
  const na2 = ddot(saRe, saRe) + ddot(saIm, saIm)
  const nb2 = ddot(sbRe, sbRe) + ddot(sbIm, sbIm)
  if (na2 <= 0 || nb2 <= 0) {
    throw new MetricError('scaled squared norm is non-positive for a nonzero state; refusing to clamp.')
  }
  const magnitude = absCx(inner)
  const value = (magnitude * magnitude) / (na2 * nb2)
  if (!Number.isFinite(value)) {
    throw new MetricError(`similarity intermediate is non-finite (${value}); refusing to clamp NaN/inf.`)
  }
  return Math.min(1, Math.max(0, value))
}

/**
 * Average pairwise coherence for a finite state system. The diagonal
 * self-similarities ARE included in the weighted average. Weights must be
 * finite and non-negative.
 */
export function systemCoherence(states: readonly ResonantState[], weights?: readonly (readonly number[])[]): number {
  const n = states.length
  if (n === 0) {
    throw new ValueError('system_coherence requires at least one state.')
  }
  if (n > MAX_STATES) {
    throw new LimitError(`system_coherence received ${n} states; limit is MAX_STATES=${MAX_STATES}.`)
  }
  const w: number[][] =
    weights === undefined
      ? Array.from({ length: n }, () => Array.from({ length: n }, () => 1))
      : weights.map((row) => row.slice())
  if (w.length !== n || w.some((row) => row.length !== n)) {
    throw new ValueError('weights must have shape (n, n).')
  }
  for (const row of w) {
    for (const weight of row) {
      if (!Number.isFinite(weight)) {
        throw new ValueError('weights must be finite.')
      }
      if (weight < 0) {
        throw new ValueError(
          'weights must be non-negative; negative weights are not meaningful for coherence averaging.',
        )
      }
    }
  }
  let total = 0
  let weightTotal = 0
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const weight = w[i]![j]!
      total += normalizedComplexSimilarity(states[i]!, states[j]!) * weight
      weightTotal += weight
    }
  }
  if (weightTotal === 0) {
    throw new ValueError('weights must not sum to zero.')
  }
  const result = total / weightTotal
  if (!Number.isFinite(result)) {
    throw new MetricError(`system_coherence produced a non-finite value (${result}).`)
  }
  return result
}
